thrust/system/cuda/detail/unique.h

```diff
@@ -794,6 +804,36 @@ unique(execution_policy<Derived> &policy,
   return cuda_cub::unique(policy, first, last, equal_to<input_type>());
 }
 
+
+template <typename BinaryPred>
+struct zip_adj_not_predicate {
+  template <typename TupleType>
+  bool __host__ __device__ operator()(TupleType&& tuple) {
+      return !binary_pred(thrust::get<0>(tuple), thrust::get<1>(tuple));
+  }
+  
+  BinaryPred binary_pred;
+};
+
+
+__thrust_exec_check_disable__
+template <class Derived,
+          class InputIt,
+          class BinaryPred>
+typename thrust::iterator_traits<InputIt>::difference_type __host__ __device__
+unique_count(execution_policy<Derived> &policy,
+       InputIt                    first,
+       InputIt                    last,
+       BinaryPred                 binary_pred)
+{
+  if (first == last) {
+    return 0;
+  }
+  auto size = last - first;
```

Comment:
This algorithm should work with forward iterators, but taking the difference of two iterators requires random access. You'll need to find a different way to implement this. @jrhemstad and @allisonvacanti suggest a possible implementation in #1619.

Reply:
I suppose thrust::distance would be appropriate then? I am having a hard time imagining how you would implement this parallel algorithm without random access iterators.

Regardless of that, I'll give `DeviceRunLengthEncode::Encode` a try, since it's simpler than I thought on my first look.

Reply:
Seems like cub doesn't like discard_iterator, since it writes and reads the first key from each run. So I think this is not possible without additional temporary storage O(size)

Reply:
Using `thrust::distance` is correct here.

Could you file a CUB issue describing the discard iterator bug?

Reply:
Turns out this has already been reported: https://github.com/NVIDIA/thrust/issues/1490

Reply:
Okay, `DeviceRunLengthEncode::Encode` works with `cub::DiscardOutputIterator`, but (at least on the Titan X I tested on) gives significantly worse performance, comparison (GB/s bandwidth)

int

| n | count_if | Encode |
|---|-------------|-----------|
| 16M |  179.173 | 109.111 |
| 1B | 254.523 | 160.037 |

long

| n | count_if | Encode |
|---|-------------|-----------|
| 16M |  259.123 | 180.335 |
| 512M | 315.969 | 215.304 |

pair<long, int>

| n | count_if | Encode |
|---|-------------|-----------|
| 16M |  351.678 284.543 |
| 256M | 392.726 | 331.611 |